### PR TITLE
Add spec.md

### LINKED ISSUE #3183: Add spec doc
**Description:**
$Subject

**Describe your task(s)**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by MohamedSabthar**
## Purpose

Fixes: [#3183](https://github.com/ballerina-platform/ballerina-standard-library/issues/3183)

## Checklist
- [x] Linked to an issue
- [ ] Updated the changelog
- [ ] Added tests

### CODE REVIEW

File: docs/spec/spec.md [hunk 1]
```diff
@@ -0,0 +1,1294 @@
+# Specification: Ballerina SerDes Library
+
+_Owners_: @MohamedSabthar @shafreenAnfar @ThisaruGuruge  
+_Reviewers_:
```
Comment (ThisaruGuruge) [inline docs/spec/spec.md]:
```suggestion
_Reviewers_: @shafreenAnfar @ThisaruGuruge    
```

File: docs/spec/spec.md [hunk 2]
```diff
@@ -0,0 +1,1294 @@
...
+
+## 1. Overview
+This specification elaborates on functionalities provided by SerDes library and how SerDes library maps the Ballerina anydata to a protocol buffer type.
+
+## 2. Proto3Schema
+`Proto3Schema` class provides APIs to perform serialization and deserialization of Ballerina anydata using protocol buffers (proto3). Following is the class definition of `Proto3Schema`
+
+```ballerina
+public class Proto3Schema {
+  *Schema;
```
Comment (ThisaruGuruge) [inline docs/spec/spec.md:65]: Let's add a section to describe the `Schema` class, and how someone can use it to write their own serializers/deserializers.

File: docs/spec/spec.md [hunk 3]
```diff
@@ -0,0 +1,1294 @@
...
+
+  // Implementation of serialize(), deserialize() functions goes here
+}
+```
+
+```ballerina
+// serdes Error
+public type Error distinct error; 
+
+// Abstract object
```
Comment (ThisaruGuruge) [inline docs/spec/spec.md]: We don't use the `Abstract Object` concept in Ballerina. This is an object type in Ballerina.

File: docs/spec/spec.md [hunk 4]
```diff
@@ -0,0 +1,1294 @@
...
+        check generateSchema(self, ballerinaDataType);
+  }
+
+  // Implementation of serialize(), deserialize() functions goes here
+}
+```
+
+```ballerina
+// serdes Error
+public type Error distinct error;
```
Comment (ThisaruGuruge) [inline docs/spec/spec.md]: Why are we including the error definition here? I think we don't need this, although describing the error type and its occurrences could be useful.

File: docs/spec/spec.md [hunk 5]
```diff
@@ -0,0 +1,1294 @@
...
+  // Implementation of serialize(), deserialize() functions goes here
+}
+```
+
+```ballerina
+// serdes Error
+public type Error distinct error; 
+
+// Abstract object
+public type Schema object {
```
Comment (ThisaruGuruge) [inline docs/spec/spec.md]: This definition should be placed earlier than the proto3Schema.

File: docs/spec/spec.md [hunk 6]
```diff
@@ -0,0 +1,1294 @@
...
+	* 3.5 [Map](#35-map)
+	* 3.6 [Table](#36-table)
+	* 3.7 [Tuple](#37-tuple)
+	* 3.8 [Enum](#38-enum)
+4. [Sample](#4-sample)
+
+## 1. Overview
+This specification elaborates on functionalities provided by SerDes library and how SerDes library maps the Ballerina anydata to a protocol buffer type.
+
+## 2. Proto3Schema
```
Comment (ThisaruGuruge) [inline docs/spec/spec.md]:
```suggestion
## 2. Proto3
```

File: docs/spec/spec.md [hunk 7]
```diff
@@ -0,0 +1,1294 @@
...
+### 2.2 `serialize` function
+Serializes the value passsed as the argument and returns `byte[]` on successful serialization or an `Error` on failure. The underlying implementation uses the previously generated proto3 message definition to serialize the provided value. Passing a value that doesn't match the type provoided during the instantiation of `Proto3Schema` object may results  in a serialization failure.
+
+### 2.3 `deserialize` function
+Deserializes the provided `byte[]` argument and returns the ballerina value with the type represented by the typedesc value provided during the `Proto3Schema` object instantiation. The underlying implementation uses the generated proto3 message definition to serialize the provided value. Passing a `byte[]` that is not a serialized value of of the specified type may result in a deserilization failure or a garbage value.
+
+## 3. Ballerina anydata to proto3 mapping
+As specified `Proto3Schema` dynamically generates proto3 message definition for given subtypes of Ballerina anydata. Following sections defines the mapping for each subtypes.
+
+
```
Comment (ThisaruGuruge) [inline docs/spec/spec.md:138]:
```suggestion
```

File: docs/spec/spec.md [hunk 8]
```diff
@@ -0,0 +1,1294 @@
...
+   string string___unionField  = 1;
+}
+</pre>
+</td>
+</tr>
+<table>
+
+
+
+## 4. Sample
```
Comment (ThisaruGuruge) [inline docs/spec/spec.md]: Let's add samples under each section when needed. We don't need a complete sample as a separate section.

File: docs/spec/spec.md (hunk 1 above)
Comment (MohamedSabthar) [reply docs/spec/spec.md]: Added

File: docs/spec/spec.md (hunk 3 above)
Comment (MohamedSabthar) [reply docs/spec/spec.md]: Comment removed

File: docs/spec/spec.md (hunk 5 above)
Comment (MohamedSabthar) [reply docs/spec/spec.md]: Moved to the new section `Schema object`

File: docs/spec/spec.md (hunk 4 above)
Comment (MohamedSabthar) [reply docs/spec/spec.md]: removed

File: docs/spec/spec.md (hunk 2 above)
Comment (MohamedSabthar) [reply docs/spec/spec.md:65]: Moved to it own section

File: docs/spec/spec.md (hunk 7 above)
Comment (MohamedSabthar) [reply docs/spec/spec.md:138]: Removed

File: docs/spec/spec.md (hunk 6 above)
Comment (MohamedSabthar) [reply docs/spec/spec.md]: the class name we used goes as `Proto3Schema`

File: docs/spec/spec.md (hunk 8 above)
Comment (MohamedSabthar) [reply docs/spec/spec.md]: +1, moved to appropriate sections

File: docs/spec/spec.md [hunk 9]
```diff
@@ -0,0 +1,1314 @@
...
+	* 4.5 [Map](#45-map)
+	* 4.6 [Table](#46-table)
+	* 4.7 [Tuple](#47-tuple)
+	* 4.8 [Enum](#48-enum)
+
+## 1. Overview
+This specification elaborates on functionalities provided by the SerDes library and how the SerDes library maps the Ballerina anydata to a protocol buffer type.
+
+## 2. Schema
+`Schema` object defines the API to perform serialization and deserialization of Ballerina anydata. One can include this `Schema` object in a class and implement their serialization and deserialization logic. The `Schema` object definition is as follows.
```
Comment (ThisaruGuruge) [inline docs/spec/spec.md]:
```suggestion
`Schema` object defines the API to perform serialization and deserialization of Ballerina anydata. You can include this `Schema` object in a class and implement your serialization and deserialization logic. The `Schema` object definition is as follows. 
```

**Review by ThisaruGuruge — APPROVED** [review]

File: docs/spec/spec.md [hunk 10]
```diff
@@ -0,0 +1,1314 @@
...
+  }
+  repeated IntOrString___UnionBuilder IntOrString___arrayField_1___unionField  = 1;
+  repeated FloatOrNill___UnionBuilder FloatOrNill___arrayField_1___unionField  = 2;
+}
+</pre>
+</td>
+</tr>
+<table>
+
+### 3.4 Record
```
Comment (ThisaruGuruge) [inline docs/spec/spec.md]:
```suggestion
### 4.4 Record
```

**Pull request merged**